walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Initiatizes return data structure and set duration parameter
+    //Collection<TimeRange> availTimes = new HashSet<>();
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+    final int length = (int) request.getDuration();
+    Collection<String>  requestAttendees = request.getAttendees();
+    
+    // Creates a List of relevant events that will restrict available time ranges
+    bookTimes = findTimeConflicts(requestAttendees, events);
+
+    // Sorts time conflict in chronological order runs the available function
+    Collections.sort(bookTimes, TimeRange.ORDER_BY_START);
+  
+    return getAvailableTimeRanges(bookTimes, length);
+  }
+
+/**Create an ArrayList of relevant events that will restrict available time ranges */
+  public ArrayList<TimeRange> findTimeConflicts(Collection<String> requestAttendees, Collection<Event> events){
+    // Set doesn't store duplicates and makes the for loop below more efficient
+    HashSet<TimeRange> bookTimes = new HashSet<TimeRange>();
+
+    // Iterates thorugh each event and decides if the TimeRange is relevant based on requestAttendees
+    for (Iterator<Event> eventIterator = events.iterator(); eventIterator.hasNext();){
+        Event event = eventIterator.next();
+      
+        for (Iterator<String> attendIterator = requestAttendees.iterator(); attendIterator.hasNext();){
+            String attendee = attendIterator.next();
+           
+            if(event.getAttendees().contains(attendee)){
+                bookTimes.add(event.getWhen());
+            }
+        }
+    }
+
+    // The times have to be returned as a list to be sorted by start time
+    ArrayList<TimeRange> bookedTimes = new ArrayList<TimeRange>(bookTimes) ; 
+    return bookedTimes;
+  }
+
+  public Collection<TimeRange> getAvailableTimeRanges(ArrayList<TimeRange> bookTimes, int length){
+    ArrayList<TimeRange> availTimes = new ArrayList<TimeRange>();
+
+    // There is a test that doesn't allow a duration longer than the day. The proper result is to return an empty list
+    if(length > TimeRange.END_OF_DAY) {
+        return availTimes;
+
+    // If there are no Time Conflicts the whole day should be returned as available
+    } else if (bookTimes.isEmpty()) {
+        availTimes.add(TimeRange.WHOLE_DAY);
+        return availTimes;
+
+    } else {
+    // Even though previous() is not used ListIterator must be used with iterating through ArrayList
+        TimeRange prevBookedTime = null;
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();) {
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(prevBookedTime != null) { 
```

Comment:
Need a comment before this with what you're doing here. E.g.,

// If we've already found a booked time, extend it

(if that's what the behavior is -- i can't really tell)

Reply:
Yeah that was a mess I was able to clean it up and I placed new comments with the refigured algorithm.